tools/autograd/gen_variable_type.py

```diff
@@ -80,6 +79,67 @@
     '_coalesced_',
 }
 
+# NOTE [ Invariant: TensorImpl and Storage Pointer Equality ]
+#
+# When a function modifies its input tensors (via inplace or out-variants),
+# it should never change the the input tensors' underlying c10::TensorImpl pointers
+# or c10::Storage pointers.
+#
+# The following code templates implement the checks for this invariant:
+SAVE_TENSOR_STORAGE = CodeTemplate("""\
+Storage ${tensor_name}_storage_saved =
+  ${tensor_name}.has_storage() ? ${tensor_name}.storage() : Storage();
+""")
+
+ENFORCE_SAME_TENSOR_STORAGE = CodeTemplate("""\
+if (${tensor_name}_storage_saved) AT_ASSERT(${tensor_name}_storage_saved.is_alias_of(${tensor_name}.storage()));
+""")
+
+SAVE_TENSORLIST_STORAGE = CodeTemplate("""\
+std::vector<Storage> ${tensorlist_name}_storage_saved(${tensorlist_name}.size());
+for (Tensor tensor : ${tensorlist_name})
+  ${tensorlist_name}_storage_saved.push_back(tensor.has_storage() ? tensor.storage() : Storage());
+""")
+
+ENFORCE_SAME_TENSORLIST_STORAGE = CodeTemplate("""\
+for (size_t i=0; i<${tensorlist_name}.size(); i++) {
+  if (${tensorlist_name}_storage_saved[i]) {
+    AT_ASSERT(${tensorlist_name}_storage_saved[i].is_alias_of(${tensorlist_name}[i].storage()));
+  }
+}
+""")
+
+SAVE_TENSOR_IMPL = CodeTemplate("""\
+c10::intrusive_ptr<TensorImpl> ${tensor_name}_impl_saved;
```

Comment:
use optional here too?

Reply:
I wasn't able to figure out how to use `c10::optional` for TensorImpl correctly here (there is always some type mismatch error going on). Also I think using `c10::intrusive_ptr` actually has pretty clear semantics, as we do want to store the pointer to the TensorImpl, and compare it when the internal function returns.